File: cpp/src/easy/two_sum.cpp
#include <vector>
#include <unordered_map>

std::vector<int> twoSum(std::vector<int>& nums, int target) {
    // map key = input vector element
    // map value = associated input vector index
    std::unordered_map<int, int> numMap;

    // loop through input vector
    for (int i = 0; i < nums.size(); ++i) {
        // if there already exists a value 'diff' in hash map, such that: diff = target - current
        if (const auto diff = numMap.find(target - nums[i]); diff != numMap.cend()) {
            // then return a vector where:
            //  first element is value (input vector index) assigned to complementary key in the map
            //  second element is the current input vector index
            return {diff->second, i};
        }
        // if no solution pair found, add:
        //  current input vector element as new map key
        //  current input vector index as that key's value
        numMap[nums[i]] = i;
    }

    throw std::runtime_error("twoSum() failed to identify solution pair");
}

File: rust/src/easy/two_sum.rs
use std::collections::HashMap;

// -----------------------------------------------------------------------------
pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let mut num_map: HashMap<i32, i32> = HashMap::new();

    for (i, &num) in nums.iter().enumerate() {
        if let Some(&diff) = num_map.get(&(target - num)) {
            return vec![diff, i as i32];
        } else {
            num_map.insert(num, i as i32);
        }
    }

    panic!("two_sum() failed to identify solution pair")
}

#[test]
pub fn two_sum_test() {
    assert_eq!(two_sum(vec![2, 7, 11, 15], 9), vec![0, 1]);
    assert_eq!(two_sum(vec![3, 2, 4], 6), vec![1, 2]);
    assert_eq!(two_sum(vec![3, 3], 6), vec![0, 1]);
}
